pmd-core/src/main/java/net/sourceforge/pmd/lang/LanguageVersionDiscoverer.java

```diff
@@ -130,9 +156,34 @@ public void setForcedVersion(LanguageVersion forceLanguageVersion) {
      */
     public List<Language> getLanguagesForFile(String fileName) {
         String extension = getExtension(fileName);
+
+        LanguageFilePattern pat = matchLanguageFilePatterns(fileName);
+        if (pat != null) {
+            // matched one of the patterns
+            for (String langId : pat.languageIds) {
+                Language lang = languageRegistry.getLanguageById(langId);
+                if (lang != null) {
+                    return Collections.singletonList(lang);
+                }
+            }
+            // language was not loaded, file is ignored.
+            return Collections.emptyList();
+        }
+
         return languageRegistry.getLanguages().stream()
                                .filter(it -> it.hasExtension(extension))
                                .collect(Collectors.toList());
+
+    }
+
+    private @Nullable LanguageFilePattern matchLanguageFilePatterns(String fileName) {
+        // match patterns from most recent to most ancient
+        for (LanguageFilePattern pat : IteratorUtil.asReversed(languageFilePatterns)) {
+            if (pat.matches(fileName)) {
+                return pat;
```

Comment:
As commented above, we should maybe return here a list of all matched patterns, so that users get a warning, if there are multiple different languages matched.